Add a status_msg field to fix issue 5150.

jagathprakash opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

- Adds a new field status_msg to taskrun spec and run spec. This field adds an user_defined / pre-defined messages for run status. 
- status_msg is used to propagate pipeline run cancellations to taskruns and tasks.

/kind bug
fixes #5150


# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Users can now differentiate if a TaskRun was cancelled by the user or by cancellation of a PipelineRun of which the TaskRun was a part of, by looking at the TaskRun's spec.StatusMessage field.
```

jagathprakash commented:
/assign lbernick

abayer commented:
/test all

abayer commented on pkg/apis/pipeline/v1alpha1/run_types.go:
```diff
@@ -60,6 +60,10 @@ type RunSpec struct {
 
+	// Status message for cancellation.
+	// +optional
+	StatusMessage RunSpecStatusMessage `json:"status_msg,omitempty"`
```
The json tag should be `statusMessage` for consistency.

jagathprakash replied:
Done.

abayer commented on pkg/apis/pipeline/v1alpha1/run_types.go:
```diff
@@ -90,6 +94,14 @@ const (
+
+const (
+	// Run cancelled as the Pipeline running it was cancelled.
+	RunCancelledByPipelineMsg RunSpecStatusMessage = "Run cancelled as pipeline has been canceled."
```
Should be `pipeline has been cancelled`. =)

lbernick replied:
Also, I think Pipeline-related constants should be defined either in the reconciler or in pipeline_types.

jagathprakash replied:
Fixed the spelling to be cancelled.

jagathprakash replied:
These constants are pipeline related but are set on a Run object and pipeline does not have a field with this object. 
As such these constants, I think, should be in run_types.go.

abayer commented on pkg/apis/pipeline/v1alpha1/run_types_test.go:
```diff
@@ -172,6 +172,27 @@ func TestRunIsCancelled(t *testing.T) {
+}
+
+func TestRunIsCancelledWithMessage(t *testing.T) {
+	expected_status_message := "test message"
```
This should be changed to `expectedStatusMessage`.

jagathprakash replied:
Done

abayer commented on pkg/apis/pipeline/v1beta1/taskrun_types.go:
```diff
@@ -52,6 +52,9 @@ type TaskRunSpec struct {
 	Status TaskRunSpecStatus `json:"status,omitempty"`
+	// Status message for cancellation.
+	// +optional
+	StatusMessage TaskRunSpecStatusMessage `json:"status_msg,omitempty"`
```
As mentioned above in the `RunSpec`, this should be `statusMessage`, not `status_msg`.

jagathprakash replied:
Done

abayer commented on pkg/apis/pipeline/v1beta1/taskrun_types.go:
```diff
@@ -90,6 +93,14 @@ const (
+
+const (
+	// TaskRun cancelled as the Pipeline running it was cancelled.
+	TaskRunCancelledByPipelineMsg TaskRunSpecStatusMessage = "TaskRun cancelled as pipeline has been canceled."
```
Again, as I noted above, `pipeline has been cancelled`, not `canceled`.

abayer commented on pkg/apis/pipeline/v1beta1/taskrun_types_test.go:
```diff
@@ -133,6 +133,27 @@ func TestTaskRunIsCancelled(t *testing.T) {
+}
+
+func TestTaskRunIsCancelledWithMessage(t *testing.T) {
+	expected_status_message := "test message"
```
`expectedStatusMessage`. =)

abayer reviewed (changes requested):
Thanks for the PR! It looks generally good, other than some naming tweaks.

abayer commented:
/retest

abayer commented:
`pull-tekton-pipeline-build-tests` failure is legit - https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5224/pull-tekton-pipeline-build-tests/1552276635263176704#1:build-log.txt%3A1538 - but it'll be fixed by renaming the field(s) to `statusMessage`.

lbernick commented on pkg/apis/pipeline/v1alpha1/run_types.go:
```diff
@@ -189,6 +201,12 @@ func (r *Run) IsCancelled() bool {
 	return r.Spec.Status == RunSpecStatusCancelled
 }
 
+// GetStatusMessage function returns the status message is one is set on the TaskRun
```
TaskRun -> Run

jagathprakash replied:
Done

lbernick commented on pkg/apis/pipeline/v1alpha1/run_types_test.go:
```diff
@@ -172,6 +172,27 @@ func TestRunIsCancelled(t *testing.T) {
+	}
+}
+
+func TestRunIsCancelledWithMessage(t *testing.T) {
```
It may be worth validating that the statusMessage is not set unless the status is set (in run_validation)

lbernick reviewed (commented):
I'm wondering which would be more useful-- a human readable sentence, or a short one word reason (e.g. "ReasonPipelineRunCancelled"). [apis.Condition](https://github.com/tektoncd/pipeline/blob/55665765e4de35b3a4fb541549ae8cdef0996641/vendor/knative.dev/pkg/apis/condition_types.go#L58-L85) has both a "Reason" and a "Message" for this purpose. Thoughts?

Also @tektoncd/core-maintainers this is an API change but it is simple enough that I'm not sure a TEP is required, WDYT?

abayer commented:
@lbernick Yeah, I don't think this requires a TEP.

lbernick commented on pkg/reconciler/taskrun/taskrun.go:
```diff
@@ -144,7 +144,7 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, tr *v1beta1.TaskRun) pkg
 	// If the TaskRun is cancelled, kill resources and update status
 	if tr.IsCancelled() {
-		message := fmt.Sprintf("TaskRun %q was cancelled", tr.Name)
+		message := fmt.Sprintf("TaskRun %q was cancelled. %s", tr.Name, tr.GetStatusMessage())
```
I think we'll want this message to be propagated to the `taskRun.status.conditions[].message`

jagathprakash replied:
@lbernick This gets propagated to taskRun.status.conditions[].message. This like does the propagation. Also test file pkg/reconciler/taskrun/taskrun_test.go has a test which checks for this propagation.

lbernick replied:
you're right! sorry I missed that.

lbernick commented on pkg/apis/pipeline/v1alpha1/run_types.go:
```diff
@@ -60,6 +60,10 @@ type RunSpec struct {
 
+	// Status message for cancellation.
+	// +optional
+	StatusMessage RunSpecStatusMessage `json:"statusMessage,omitempty"`
```
could you add some info to docs/runs.md about how custom task controllers should support this message?

jagathprakash replied:
Added some info in runs.md on how statusMessage can be used by the custom controller developer.

jagathprakash commented:
@abayer Thanks for the review. Modified the test/cancel_test.go to do an E2E test on this feature.

abayer commented on test/cancel_test.go:
```diff
@@ -145,6 +145,13 @@ spec:
+					t.Fatalf("Status is %s while it should have been %s", taskrunItem.Spec.Status, v1beta1.TaskRunSpecStatusCancelled)
+				}
+				if taskrunItem.Spec.StatusMessage != v1beta1.TaskRunCancelledByPipelineMsg {
+					t.Fatalf("Status message is set to %s while it should be %s.", taskrunItem.Spec.StatusMessage, v1beta1.TaskRunCancelledByPipelineMsg)
```
👍

abayer commented on pkg/apis/pipeline/v1alpha1/run_types.go:
```diff
@@ -90,6 +94,14 @@ const (
+type RunSpecStatusMessage string
+
+const (
+	// Run cancelled as the Pipeline running it was cancelled.
```
Please give this a valid godoc comment - i.e., one starting with `// RunCancelledByPipelineMsg ...`

jagathprakash replied:
Done.

abayer commented on pkg/apis/pipeline/v1alpha1/run_types.go:
```diff
@@ -189,6 +201,12 @@ func (r *Run) IsCancelled() bool {
 
+// GetStatusMessage function returns the status message is one is set on the Run
+// If nothing is set as status message, GetStatusMessage returns an empty string.
+func (tr *Run) GetStatusMessage() RunSpecStatusMessage {
```
Current `golangci-lint` versions will start barfing on the receiver name here not matching others in the file - so please change this to `func (r *Run) ...`.

jagathprakash replied:
Done, thanks for pointing that out.

abayer commented on pkg/apis/pipeline/v1beta1/taskrun_types.go:
```diff
@@ -90,6 +93,14 @@ const (
+type TaskRunSpecStatusMessage string
+
+const (
+	// TaskRun cancelled as the Pipeline running it was cancelled.
```
Please give this a valid godoc comment - i.e., one starting with `// TaskRunCancelledByPipelineMsg ...`

jagathprakash replied:
Done

abayer commented on pkg/apis/pipeline/v1alpha1/run_types.go:
```diff
@@ -90,6 +94,14 @@ const (
+
+const (
+	// Run cancelled as the Pipeline running it was cancelled.
+	RunCancelledByPipelineMsg RunSpecStatusMessage = "Run cancelled as pipeline has been cancelled."
```
I'd also reword this (and the `TaskRun` equivalent) to something like `Run cancelled as the PipelineRun it belongs to has been cancelled.`

jagathprakash replied:
Done.

abayer commented:
/test pull-tekton-pipeline-alpha-integration-tests

abayer commented:
/test all

abayer commented:
FYI, I've just changed the title of this PR to have `WIP: ` and marked it as ready for review, so that CI fires. Prow doesn't automatically run presubmit jobs for draft PRs. =)

abayer commented:
`alpha-integration` failure is a flake, but `build-tests` failure is legit - https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5224/pull-tekton-pipeline-build-tests/1554476253346533376#1:build-log.txt%3A1649 - you really just need to run `hack/update-codegen.sh`, since that calls `hack/update-openapigen.sh`.

jagathprakash commented on pkg/apis/pipeline/v1beta1/taskrun_types.go:
```diff
@@ -52,6 +52,9 @@ type TaskRunSpec struct {
 	Status TaskRunSpecStatus `json:"status,omitempty"`
+	// Status message for cancellation.
+	// +optional
+	StatusMessage TaskRunSpecStatusMessage `json:"status_msg,omitempty"`
```
The v1 TaskRun is not yet implemented. I will add this in a separate PR.

jagathprakash commented:
> FYI, I've just changed the title of this PR to have `WIP: ` and marked it as ready for review, so that CI fires. Prow doesn't automatically run presubmit jobs for draft PRs. =)

Thanks!

jagathprakash commented:
> `alpha-integration` failure is a flake, but `build-tests` failure is legit - https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5224/pull-tekton-pipeline-build-tests/1554476253346533376#1:build-log.txt%3A1649 - you really just need to run `hack/update-codegen.sh`, since that calls `hack/update-openapigen.sh`.

@abayer hack/update-codegen.sh did not generate any files or changes. How do I fix this?

jagathprakash commented:
/test pull-tekton-pipeline-build-tests

jagathprakash commented:
/test pull-tekton-pipeline-alpha-integration-tests

abayer commented:
> hack/update-codegen.sh did not generate any files or changes. How do I fix this?

You might as well try running `hack/update-openapigen.sh` directly - that's where the issue is. When I run it locally, I get changes to `pkg/apis/pipeline/v1beta1/openapi_generated.go` and `pkg/apis/pipeline/v1beta1/swagger.json`.

jagathprakash commented:
/test 

> Thanks for the PR! It looks generally good, other than some naming tweaks.
> 
> Oh, and I'd like to see this tested in the end-to-end tests, probably in `test/cancel_test.go`.

Added a test in test/cancel_test.go and also addressed all naming issues.

jagathprakash commented:
> > hack/update-codegen.sh did not generate any files or changes. How do I fix this?
> 
> You might as well try running `hack/update-openapigen.sh` directly - that's where the issue is. When I run it locally, I get changes to `pkg/apis/pipeline/v1beta1/openapi_generated.go` and `pkg/apis/pipeline/v1beta1/swagger.json`.

Thanks! That worked, running all the tests now.

abayer commented:
The unit test failure is a flake, the build test failure is a lint issue - https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5224/pull-tekton-pipeline-build-tests/1554577871807188992#1:build-log.txt%3A1648

jagathprakash commented:
/test pull-tekton-pipeline-build-tests

abayer commented on docs/runs.md:
```diff
@@ -192,7 +192,11 @@ Supporting timeouts is optional but recommended.
    the waiting timer, tear down the approval listener).
-4. Once resources or timers are cleaned up it is good practice to set a
+4. A Custom Task author can differentiate between a run cancelled by an user from a
+   run cancel intiated by a pipeline run cancel by looking at the Run.Spec.StatusMessage.
```
A few formatting/spelling notes:
* I'd go with "... a `Run` ..." rather than "... a run ..."
* "Run.Spec.StatusMessage" should be in backticks.
* "intiated" should be "initiated". =) And I'd phrase that as "a `Run` cancellation initiated by..."

jagathprakash replied:
Done.
Also made some more changes to the doc on similar lines.

jagathprakash commented:
> The unit test failure is a flake, the build test failure is a lint issue - https://prow.tekton.dev/view/gs/tekton-prow/pr-logs/pull/tektoncd_pipeline/5224/pull-tekton-pipeline-build-tests/1554577871807188992#1:build-log.txt%3A1648

Thanks for helping find this!

abayer commented:
@jagathprakash Please add a release note in the PR description - and in the `release-note` block, remove the space between the three backticks and `release-note`. Thanks!

abayer commented:
/retest

jagathprakash commented:
> @jagathprakash Please add a release note in the PR description - and in the `release-note` block, remove the space between the three backticks and `release-note`. Thanks!

Done. Thanks for the tip!

lbernick commented on pkg/apis/pipeline/v1alpha1/run_types.go:
```diff
@@ -189,6 +202,12 @@ func (r *Run) IsCancelled() bool {
 
+// GetStatusMessage function returns the status message is one is set on the Run
+// If nothing is set as status message, GetStatusMessage returns an empty string.
+func (r *Run) GetStatusMessage() RunSpecStatusMessage {
```
does this function need to exist?

jagathprakash replied:
Done. Removed the function.

lbernick commented on pkg/apis/pipeline/v1beta1/taskrun_types.go:
```diff
@@ -424,6 +436,14 @@ func (tr *TaskRun) IsCancelled() bool {
 
+// GetStatusMessage function returns the status message is one is set on the TaskRun
+// If nothing is set as status message, GetStatusMessage returns an empty string.
+func (tr *TaskRun) GetStatusMessage() TaskRunSpecStatusMessage {
```
same comment here, I don't think this function is necessary

jagathprakash replied:
Done. Removed the function.

lbernick reviewed (commented):
Looks great! just a nit but will lgtm once it is resolved

lbernick commented:
/lgtm

jagathprakash commented:
Thanks @abayer and @lbernick for the detailed reviews!
